compiler/rustc_middle/src/ty/adt.rs

```diff
@@ -103,6 +104,8 @@ pub struct AdtDefData {
     flags: AdtFlags,
     /// Repr options provided by the user.
     repr: ReprOptions,
+    /// Target features that functions taking objects of this type by value will enable.
+    target_features: Vec<Symbol>,
```

Comment:
I suppose you dont intend to support `struct Foo<T>(T);` where `Foo` is provided a type argument of a target feature struct, e.g. `Foo<Avx512>`, automatically enabling the target feature on the function. If you did want to support anything like that eventually this would be fundamentally the wrong impl design I think

not really an issue if this is as far as the feature will go wrt determining if a value is passed to a function though :3

Reply:
I think a `Foo<T>(T)` with `T = Avx512` should behave the same as a `FooAvx512(Avx512)`, or it would be very surprising - how would this design be broken in this case and what do you think would be a better option?

Reply:
Ah I think the comment confused me a bit. `struct Foo<T>(T)` wouldnt have any target features defined on it but passing a `Foo<Avx512>` to a function would enable the target feature. The comment would seem to imply that it's the complete set of target features. In practice I suppose you need to store in the adt def what target feature attrs were applied *anyway* even if everywhere else you walk the fields. Seems fine to me then ^^

Reply:
Yup, indeed - other than generics, a reason to walk the type and not store the complete set of features in the ADT is that types that *enable* new features ought to behave differently (in particular, be unsafe to construct).

I will update the comment :-)